File: server/routes/notes.js
const express = require('express');
const router = express.Router();
const Note = require('../models/Note');
const { protect } = require('../middleware/auth');

router.use(protect);

// @route   GET /api/notes
// @desc    Get all notes for logged in user
// @access  Private
router.get('/', async (req, res) => {
    try {
        const notes = await Note.find({ user: req.user._id }).sort({ updatedAt: -1 });
        res.json(notes);
    } catch (error) {
        console.error('Get notes error:', error);
        res.status(500).json({ message: 'Server error' });
    }
});

// @route   POST /api/notes
// @desc    Create a new note
// @access  Private
router.post('/', async (req, res) => {
    try {
        const note = await Note.create({
            ...req.body,
            user: req.user._id,
        });
        res.status(201).json(note);
    } catch (error) {
        console.error('Create note error:', error);
        res.status(500).json({ message: 'Server error' });
    }
});

// @route   PUT /api/notes/:id
// @desc    Update a note
// @access  Private
router.put('/:id', async (req, res) => {
    try {
        const note = await Note.findById(req.params.id);

        if (!note) {
            return res.status(404).json({ message: 'Note not found' });
        }

        if (note.user.toString() !== req.user._id.toString()) {
            return res.status(403).json({ message: 'Not authorized to update this note' });
        }

        const updatedNote = await Note.findByIdAndUpdate(
            req.params.id,
            req.body,
            { new: true, runValidators: true }
        );

        res.json(updatedNote);
    } catch (error) {
        console.error('Update note error:', error);
        res.status(500).json({ message: 'Server error' });
    }
});

// @route   DELETE /api/notes/:id
// @desc    Delete a note
// @access  Private
router.delete('/:id', async (req, res) => {
    try {
        const note = await Note.findById(req.params.id);

        if (!note) {
            return res.status(404).json({ message: 'Note not found' });
        }

        if (note.user.toString() !== req.user._id.toString()) {
            return res.status(403).json({ message: 'Not authorized to delete this note' });
        }

        await Note.findByIdAndDelete(req.params.id);
        res.json({ message: 'Note deleted successfully' });
    } catch (error) {
        console.error('Delete note error:', error);
        res.status(500).json({ message: 'Server error' });
    }
});

module.exports = router;

File: server/routers/notes.py
from fastapi import APIRouter, HTTPException, status, Depends
from typing import List
from bson import ObjectId
from datetime import datetime

from database import get_database
from models import NoteCreate, NoteUpdate, NoteResponse, UserInDB
from auth import get_current_user

router = APIRouter(prefix="/api/notes", tags=["notes"])


@router.get("/", response_model=List[NoteResponse])
async def get_notes(current_user: UserInDB = Depends(get_current_user)):
    """Get all notes for the current user"""
    db = get_database()
    notes = await db.notes.find({"user": current_user.id}).sort("updatedAt", -1).to_list(length=None)
    
    return [NoteResponse(_id=str(note["_id"]), **{k: v for k, v in note.items() if k != "_id" and k != "user"}) for note in notes]


@router.post("/", response_model=NoteResponse, status_code=status.HTTP_201_CREATED)
async def create_note(note_data: NoteCreate, current_user: UserInDB = Depends(get_current_user)):
    """Create a new note"""
    db = get_database()
    
    note_dict = note_data.model_dump()
    note_dict["user"] = current_user.id
    note_dict["createdAt"] = datetime.utcnow()
    note_dict["updatedAt"] = datetime.utcnow()
    
    result = await db.notes.insert_one(note_dict)
    created_note = await db.notes.find_one({"_id": result.inserted_id})
    
    return NoteResponse(_id=str(created_note["_id"]), **{k: v for k, v in created_note.items() if k != "_id" and k != "user"})


@router.put("/{note_id}", response_model=NoteResponse)
async def update_note(note_id: str, note_data: NoteUpdate, current_user: UserInDB = Depends(get_current_user)):
    """Update a note"""
    db = get_database()
    
    note = await db.notes.find_one({"_id": ObjectId(note_id), "user": current_user.id})
    if not note:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Note not found")
    
    update_data = {k: v for k, v in note_data.model_dump(exclude_unset=True).items() if v is not None}
    if update_data:
        update_data["updatedAt"] = datetime.utcnow()
        await db.notes.update_one({"_id": ObjectId(note_id)}, {"$set": update_data})
    
    updated_note = await db.notes.find_one({"_id": ObjectId(note_id)})
    return NoteResponse(_id=str(updated_note["_id"]), **{k: v for k, v in updated_note.items() if k != "_id" and k != "user"})


@router.delete("/{note_id}")
async def delete_note(note_id: str, current_user: UserInDB = Depends(get_current_user)):
    """Delete a note"""
    db = get_database()
    
    note = await db.notes.find_one({"_id": ObjectId(note_id), "user": current_user.id})
    if not note:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Note not found")
    
    await db.notes.delete_one({"_id": ObjectId(note_id)})
    return {"message": "Note deleted successfully"}
